vcs/github/github.go

```diff
@@ -299,16 +304,90 @@ func (w *workingCopy) checkout(ctx context.Context, version vcs.VersionNumber) e
 
 		return err
 	}
-	w.version = version
-	return nil
+	for {
+		err := w.g.git(ctx, w.dir, nil, "checkout", string(version))
+		if err == nil {
+			w.version = version
+			return nil
+		}
+		var exitErr *exec.ExitError
+		if !errors.As(err, &exitErr) {
+			return handleVersionNotFound(err)
+		}
+		if exitErr.ExitCode() != 128 {
+			return handleVersionNotFound(err)
+		}
+		tries += 1
+		if tries > maxTries {
+			return fmt.Errorf("git checkout exited with a 128 error code %d times in a row, aborting (%w)", maxTries, err)
+		}
+		select {
+		case <-ctx.Done():
+			return err
+		case <-time.After(100 * time.Millisecond):
+			w.g.config.Logger.Debug(ctx, "git checkout exited with a 128 error code, retrying...")
+		}
+	}
 }
 
 func (w *workingCopy) reset(ctx context.Context) error {
-	return w.g.git(ctx, w.dir, nil, "reset", "--hard")
+	// We are implementing a retry here because git seems to have issues on GitHub Actions, possibly due to a race
+	// condition or inadequate cleanup.
+	tries := 0
+	const maxTries = 10
+	for {
+		err := w.g.git(ctx, w.dir, nil, "reset", "--hard")
+		if err == nil {
+			return nil
+		}
+		var exitErr *exec.ExitError
+		if !errors.As(err, &exitErr) {
+			return err
+		}
+		if exitErr.ExitCode() != 128 {
+			return err
+		}
+		tries += 1
+		if tries > maxTries {
+			return fmt.Errorf("git reset --hard exited with a 128 error code %d times in a row, aborting (%w)", maxTries, err)
+		}
+		select {
+		case <-ctx.Done():
+			return err
+		case <-time.After(100 * time.Millisecond):
+			w.g.config.Logger.Debug(ctx, "git reset --hard exited with a 128 error code, retrying...")
+		}
+	}
 }
 
 func (w *workingCopy) clean(ctx context.Context) error {
-	return w.g.git(ctx, w.dir, nil, "clean", "-fd")
+	// We are implementing a retry here because git seems to have issues on GitHub Actions, possibly due to a race
+	// condition or inadequate cleanup.
+	tries := 0
+	const maxTries = 10
+	for {
```

Comment:
This retry logic could be separated and reused. I would prefer it that way to ensure all the retries behave the same way for different commands.